### DISCUSSION TITLE
making an extracted quotations manager within zotero+bbt

### DISCUSSION THREAD

**Discussion by ezrajteboul** [retorquere/zotero-better-bibtex#1798] [Q&A] [answered]
Hi, as suggested there, I'm migrating a [thread ](https://forums.zotero.org/discussion/comment/381807#Comment_381807)from the zotero forums over to the bbt github so we can use markdown. 

I'm trying to code an annotation scanner that will compile all the quotes and comments in a note that are flagged by a # and will produce a .tex file that contains the corresponding quotes, proper \cite commands, and comments, ordered by year of publication.

eg, if I have, across a number of other unrelated children notes with extracted annotations and other comments: 

>"In other words, the artists attempted a kind of "reverse engineering" of traditional conceptions of community formation." (Dewar 2009:127)
>
> #reverseeng a reverse engineering of community through the development of capital's folk music (note on p.127) 

 in one note, for the parent item with the bibkey "Dewar2009," and, in another note for the parent item with the bibkey "Gell1998" : 

> "Style, I argue, is 'relations between relations' of forms." (Gell 1998:237)
> 
> #reverseeng good definition of style (note on p.237) 

I would like the desired code to output, when ran for `#reverseeng`, to output something like 
```
\documentclass{article}
\usepackage[utf8]{inputenc}
\usepackage[english]{babel}

\usepackage{biblatex}
\addbibresource{library.bib}

\begin{document}

``Style, I argue, is 'relations between relations' of forms.'' \cite[237]{gell1998}

good definition of style

``In other words, the artists attempted a kind of ``reverse engineering"'' of traditional conceptions of community formation.'' \cite[127]{Dewar2009}

 a reverse engineering of community through the development of capital's folk music

\printbibliography

\end{document}
```
Now, as emilianoeheyns pointed out on the zotero forum, this is more complicated than I anticipated. I do understand that html to tex is not easy. Would it be easier to simply output something closer to this? 
```
``Style, I argue, is 'relations between relations' of forms.'' \cite[237]{gell1998}

good definition of style

``In other words, the artists attempted a kind of ``reverse engineering"'' of traditional conceptions of community formation.'' \cite[127]{Dewar2009}

 a reverse engineering of community through the development of capital's folk music
```

I don't anticipate running this script or translator often, so I can add the document setup and end matter by hand every time - but it would be the point to at least generate text like that in the second code snippet, with just the proper cite keys and the quotes + comments arranged by year. In other words, this barely has to output any .tex code, it can be mostly plaintext with just \cite[]{} commands where appropriate.

Thanks for your help / thinking about this! 

all the best, 

Ezra

**Comment by retorquere** [comment, marked as answer]
I'm emilianoeheyns.

The sections that you put in blockquote, those are the entirety of the note? There are no multiple of these annotations in a single Zotero note?

Are these annotations always structured exactly the same? Always exactly

```
#<single-word keyword> <unstructured text> (note on p.<number>)
```

There isn't any difference in complexity between the two outputs shown above; the problem is not in the surrounding text, but in translating the Zotero note bodies, which can contain markup and html-entities. Unfortunately, pandoc does not convert citations in html to citations in tex. So that leaves either converting HTML markup to tex markup in your code, or stripping HTML markup and exporting plain-text tex without markup.

**Reply by ezrajteboul** [reply]
Hi!  I appreciate your really prompt replies (also, that you maintain better bibtex!!!! writing my dissertation would have been a lot less fun without it). your questions are in italics to differentiate them from blockquotes: 

_The sections that you put in blockquote, those are the entirety of the note?_ 

No, the note would contain other extracted quotes and comments, potentially with other #flags. Some quotes don't have comments or flags, some might have just flags. here is an example of the above blockquote with its surrounding context: 

> "Style, I argue, is 'relations between relations' of forms." (Gell 1998:237)
>
> #reverseeng good definition of style (note on p.237)
> 
> "Artworks arc like social agents, in that they are the outcome of social initiatives which reflect a specific, socia11y inculcated 
> sensibility." (Gell 1998:242)
> 
> #agency good definition of artworks and agency (note on p.242)
>
> "Personal agency, as intervention in the causal milieu, generates one of these 'distributed objects', that is, all the material 
> differences in 'the way things are' from which some particular agency can be abducted." (Gell 1998:245)

_There are no multiple of these annotations in a single Zotero note?_

There is almost always more than one carriage-return-separated quote in a note, sometimes multiple different #flags in a note (as above, with #reverseeng and #agency), and sometimes multiple instances of the same #flag in the same note. I can give you a longer example containing all this if you want. 

_Are these annotations always structured exactly the same? Always exactly_
` #<single-word keyword> <unstructured text> (note on p.<number>)`

yes, although with the comment it would be something more like: 
``` 
"<unstructured quote text>"

 #<single-word keyword> <unstructured comment text> (note on p.<number>)
```
These are obtained by running the "extract annotations" function from the zotero item right-click dropdown. I believe this function is provided by the zotfile add-on. 

_stripping HTML markup and exporting plain-text tex without markup._

If I understand this correctly that would be fine, as long as there's a `\cite[<number>]{proper bibkey}` at the end of every quote, where number is the corresponding page number from the above `note on p.<number>` for the corresponding quote. 

Is this best done with a translator, or with something else?

**Reply by retorquere** [reply]
> No, the note would contain other extracted quotes and comments, potentially with other #flags. Some quotes don't have comments or flags, some might have just flags. here is an example of the above blockquote with its surrounding context:

You can select text and select "quote reply" from the triple-dot menu above the comment BTW.

So yeah, you'd have to figure out how to parse this, since the notes may look like plain text, they're HTML, and linebreaks are not `\n` characters but `<br>` or `</p>`.

> If I understand this correctly that would be fine, as long as there's a `\cite[<number>]{proper bibkey}` at the end of every quote, where number is the corresponding page number from the above `note on p.<number>` for the corresponding quote.

That's up to you to generate. You can output whatever you want at this point.

In the translator you can just get `item.citationKey`, in plugin code you can use `item.getField('citationKey')` with BBT 5.4.1.

> Is this best done with a translator, or with something else?

I would say a translator would be easiest for you to maintain. The javascript runner you'd have to copy-paste anytime you used it, but to introduce the search term in the translator you'd have to enter it into a preference since translators can't get other outside information. Last option would be a full blown plugin but I'd say that is way too much effort for what you're trying to do.

**Reply by ezrajteboul** [reply]
Thank you for the extra details! I'm going to look into text parsing with javascript + translators and I'll come back to this forum when I have further technical questions...

Ezra
